Add option to dromedary.py to generate Opam switch and install packages

Release-Candidate opened the pull request:
Add the possibility to create a new Opam switch at a given path or with a given name, setting its compiler and install packages in this newly created switch. This Opam switch's packages are used to generate the `BUCK` file.

Example configuration:

```json
{
    "name": "./third-party",
    "compiler": "ocaml-variants",
    "packages": [
        "menhirLib",
        "sedlex=3.2",
        "alcotest>=1.7.0"
    ]
}
```

vsiles commented:
Thank you for the PR.
I need to think about it a bit, but overall I think it is a good idea. I'll get back to you soon

vsiles commented:
Sorry I got sidetracked this week.
I think it's a good idea. Will try to review asap

vsiles commented on dromedary.py:
```diff
@@ -163,6 +180,143 @@ def print_output(out: subprocess.CompletedProcess) -> None:
+        config_file (str): The path to the JSON config file to use.
+
+    Returns:
+        Tuple[str, Dict[str, Any]]: The name of the generated Opam switch and
```
"The name of" vs a Tuple. I think this comment needs some update.

Release-Candidate replied:
Oh, I see, I'm going to change the phrasing a bit.

vsiles commented on dromedary.py:
```diff
@@ -163,6 +180,143 @@ def print_output(out: subprocess.CompletedProcess) -> None:
+    if gen_switch_name is None:
+        gen_switch_name = DEFAULT_OPAM_SWITCH_NAME
+
+    if gen_switch_name.startswith("."):
```
Be careful of `..`

Release-Candidate replied:
I've chosen `.`, so that it works with `./`, `../` and hidden directories like `.NAME`, which is what Opam interprets as a relative path (I guess they use the same test).
 
I guess I'll add a comment ;)

vsiles replied:
I was not aware of that :) TIL !

vsiles commented on dromedary.py:
```diff
@@ -60,6 +69,14 @@
 `OPAM_SWITCH_ENV_CMD` containing the name of the current or given opam
 switch."""
 
+DEFAULT_OPAM_SWITCH_NAME = "./"
```
Reading the rest of the code and coming back here: this feels strange as a "name". It looks like you might end up calling `opam switch create` with the name set as `"./"` which doesn't seem like a good idea. Can you explain a bit more what would happen (I might be wrong), or update this logic in a more clear way (you are allowed to choose a default name like `generated-switch-for-buck2`, or whatever)

Release-Candidate replied:
The problem is the naming, Opam calls this `plain name` or `directory name` or just `name`. This argument to `opam switch create` can be a directory or a name. If it is a directory (relative or absolute), a "sandbox" is created in this directory, if it is a name, a global switch with that name is created (in `~/.opam/`) but only, if a compiler is given too or `name` is the "name" of a compiler. What exactly the "name" of a compiler is also a bit complicated ;)
The idea is that the default is creating a sandbox in the directory the config file is located in, as the "normal" usage is `opam switch create ./ ocaml-base-compiler` in the project's root. 

I did think about naming it `name_or_path`, but I decided to stay "close" to Opam's usage.

vsiles replied:
👍

vsiles commented on dromedary.py:
```diff
@@ -163,6 +180,143 @@ def print_output(out: subprocess.CompletedProcess) -> None:
+    print(
+        f"Generating Opam switch '{gen_switch_name}' using compiler '{gen_switch_compiler}'"
+    )
+    if switch_config.get("packages") is None:
```
You are doing this check at multiple locations. I would implement a "check" logic right after the call to `read_json` to sanitize the content, so you avoid dealing with missing / empty fields everywhere.

Release-Candidate replied:
That's a great idea, will do.

vsiles reviewed (changes requested):
Can you clarify how you deal with missing switch names exactly, and factorize the "correctness" logic of your input ?

vsiles commented on dromedary.py:
```diff
@@ -163,6 +180,166 @@ def print_output(out: subprocess.CompletedProcess) -> None:
 
 
+###############################################################################
+def run_cmd_output(cmd_args: List[Any], cmd_env: Optional[Dict[str, str]]) -> str:
```
Why is `cmd_args` a `List[Any]` instead of `List[str]` ?

Release-Candidate replied:
I have to put quotes (because of whitespace in paths and e.g. a package may contain a "<" or ">") around each argument which implicitly converts (well, calls `str()`) each argument to a string. This just makes that explicit. If you prefer a `List[str]`, I can of course change that. Or I could add a  [Protocol](https://docs.python.org/3/library/typing.html#typing.Protocol)  which made that explicit and use that instead of `Any`.

vsiles commented on dromedary.py:
```diff
@@ -163,6 +180,166 @@ def print_output(out: subprocess.CompletedProcess) -> None:
+    """
+    std_out: str = ""
+    with subprocess.Popen(
+        " ".join(map(lambda e: f"'{str(e)}'", cmd_args)),
```
Feedback from @grievejia :

If cmd_args is a list of string, then this is just `" ".join(cmd_args)`.
Taking a step back though -- I'm not really sure how much value this big block of code brings to the table, compared to just doing a simple `subprocess.run(cmd_args)`. If we really don't like the buffered capture behavior, then it can be disabled by simple set `bufsize=1` which imo is far more straightforward than all the `poll()` that's happening here.
Also is there a strong reason why stdout and stderr are merged into a single string?

Release-Candidate replied:
It can never be just a `" ".join(cmd_args)`, as each argument has to be quoted, so it is at least a `" ".join(map(lambda e: f"'{e}'", cmd_args))`. Which does `str(e)` (call `e.__str__()`, even for strings). 
I'm going to add that to the function's documentation.

Setting `bufize=1` means reading one line after the other. This does not work if the command does some kind of "fancy" progress display involving e.g. backspace. Which Opam does not do as of now, so it would be safe to use until Opam changes that, if you prefer that. 
As the switch creation and package installation can take quite some time, using fully buffered output is not an option.

The reason to merge stdout and stderr is to not have to do the equivalent of a `select` on `stdout` and `stderr` in Python, as doing the "naive" version of just reading from stderr too is going to block forever, if no data is waiting in one of the pipes.

Release-Candidate replied:
By not parsing the name of the switch any more, there is no need for pipes, so I can use a single  `subprocess.run` without pipes, which solves all the above problems.

vsiles commented on dromedary.py:
```diff
@@ -163,6 +180,166 @@ def print_output(out: subprocess.CompletedProcess) -> None:
+    return std_out
+
+
+class SwitchConfig(NamedTuple):
```
Feedback from @grievejia 👍 
I would recommend frozen dataclass over named tuple for Python >= 3.7

Release-Candidate replied:
I personally dislike decorators, but will change that.

vsiles commented on dromedary.py:
```diff
@@ -163,6 +180,166 @@ def print_output(out: subprocess.CompletedProcess) -> None:
+        config_file (str): The path to the JSON config file to use.
+
+    Returns:
+        SwitchConfig: A `Tuple` containing the name of the generated Opam
```
The return type and the return value does not seem to match this docstring.

Release-Candidate replied:
Oh, yes.

vsiles commented on dromedary.py:
```diff
@@ -163,6 +180,166 @@ def print_output(out: subprocess.CompletedProcess) -> None:
+        SwitchConfig: A `Tuple` containing the name of the generated Opam
+                      switch and the list of packages to install.
+    """
+    print(f"Using config file '{config_file}'")
```
Feedback from @grievejia 

I believe the usual convention is that human-oriented helper messages like this go to stderr, whereas stdout is strictly reserved for the the "proper" output of the script? Same for all other occurrences of print() in this diff.

Release-Candidate replied:
The idea behind that convention is to keep messages for the user separate from the output of the program that is, for example, piped into another program. But `dromedary.py` does not produce data that is meant to be parsed or piped into another program at all, that's what its generated output file and the optionally created intermediate JSON file are for. So, instead of making the script less usable for a human user by printing non-errors to stderr, I rather add an option to output machine-readable data.

vsiles commented on dromedary.py:
```diff
@@ -163,6 +180,166 @@ def print_output(out: subprocess.CompletedProcess) -> None:
+    match = re.search(switch_name_rex, std, re.UNICODE | re.DOTALL | re.MULTILINE)
+    if match is None:
+        print(
+            f"Error: could not successfully parse the output of 'opam switch create':\n{std}\nUsing regex {switch_name_rex}",
```
Feedback from @grievejia  and I fully agree (sorry I missed it):

Potentially silly question: what's the point of parsing the output of opam switch create to extract the switch name? Don't we know the name of the switch (valid_config.name) already?
This feels rather fragile as it relies on opam keeping its human-oriented output format stable, which imo is not very likely.

Release-Candidate replied:
As Opam's handling of switch names (which can or must be "compiler names") and the differentiation between paths and "real" names is complex, I do not trust Opam to use the same "name" as it has been given (it happened once to me that the name I passed had not been the same as the one Opam actually generated). 
But again thinking about it, it is better to try and use the "wrong" name and fail when calling `opam env`, than trying to go on using the "actual" switch name. As I actually never liked that part either, let's remove it.

vsiles reviewed (changes requested):
Some more feedback after internal review @ Meta.

Release-Candidate commented:
I've addressed everything except for the `List[Any]` and the printing of messages to stdout.
Also added a flag to always answer question with "yes" to `opam install`.

grievejia commented:
Many thanks for the update @Release-Candidate ! Will stamp this after @vsiles import the latest change again.

Release-Candidate commented:
You're welcome!
Thank you @grievejia and especially @vsiles for helping me adding `dromedary.py` to ocaml-scripts!
